publisher/htmlElementsCollector.go

```diff
@@ -86,109 +126,91 @@ func (w *cssClassCollectorWriter) Write(p []byte) (n int, err error) {
 		}
 
 		if w.isCollecting {
+			// if is collecting, cycle through byte stream until end bracket ">" is found
+			// disregard any ">" if within a quote
+			// write bytes until found to buffer
 			for ; i < len(p); i++ {
 				b := p[i]
 				w.toggleIfQuote(b)
+				w.buff.WriteByte(b)
+
 				if !w.inQuote && b == '>' {
 					w.endCollecting()
 					break
 				}
-				w.buff.WriteByte(b)
 			}
+		}
 
-			if !w.isCollecting {
-				if w.inPreTag != "" {
-					s := w.buff.String()
-					if tagName, isEnd := w.parseEndTag(s); isEnd && w.inPreTag == tagName {
-						w.inPreTag = ""
-					}
-					w.buff.Reset()
-					continue
-				}
-
-				// First check if we have processed this element before.
-				w.collector.mu.RLock()
-
-				// See https://github.com/dominikh/go-tools/issues/723
-				//lint:ignore S1030 This construct avoids memory allocation for the string.
-				seen := w.collector.elementSet[string(w.buff.Bytes())]
-				w.collector.mu.RUnlock()
-				if seen {
-					w.buff.Reset()
-					continue
-				}
-
-				s := w.buff.String()
-
-				w.buff.Reset()
+		// if no end bracket ">" is found while collecting, but the stream ended
+		// this could mean we received chunks of a stream from e.g. the minify functionality
+		// next if loop will be skipped
 
-				if strings.HasPrefix(s, "</") {
-					continue
+		// at this point we have collected an element line between angle brackets "<" and ">"
+		if !w.isCollecting {
+			s := w.buff.String()
+			w.buff.Reset()
+
+			// filter out unwanted tags
+			// empty string, just in case
+			// if within preformatted code blocks <pre>, <textarea>, <script>, <style>
+			// comments and doctype tags
+			// end tags
+			switch {
+			case s == "": // empty string
+				continue
+			case w.inPreTag != "": // within preformatted code block
+				if tagName, isEnd := parseEndTag(s); isEnd && w.inPreTag == tagName {
+					w.inPreTag = ""
 				}
+				continue
+			case strings.HasPrefix(s, "<!"): // comment or doctype tag
+				continue
+			case strings.HasPrefix(s, "</"): // end tag
+				continue
+			}
 
-				key := s
-
-				s, tagName := w.insertStandinHTMLElement(s)
-				el := parseHTMLElement(s)
-				el.Tag = tagName
-				if w.isPreFormatted(tagName) {
-					w.inPreTag = tagName
-				}
+			// check if we have processed this element before.
+			w.collector.mu.RLock()
+			// See https://github.com/dominikh/go-tools/issues/723
+			//lint:ignore S1030 This construct avoids memory allocation for the string.
+			seen := w.collector.elementSet[s]
```

Comment:
With this revamped implementation, the comments surrounding it does not make sense.

This relates to my benchmark comments, but I'm pretty sure we need to preserve the old behaviour.

Reply:
I removed the comment. I don't think we need the old behavior, as further up the buffer is written into a string and reset. So here we just have the string left to compare it to the elementSet list.